Project-Management/CoreTeam/workspace-notes.md

```diff
@@ -0,0 +1,43 @@
+# Workspace
+
+## Issues with the current workspace
+
+Focussing mainly on `MatrixWorkspace`.
+
+1. Constructed iteratively, i.e., workspaces start their life in an incomplete and dysfunctional state, vital information is added only later (size, instrument, spectrum to detector mapping).
+1. No encapsulation for data it contains, thus violating invariants:
+   - Access to X, Y, and E makes breaking or abusing internal data easy.
```

Comment:
Sometimes I feel like we're missing basic mid-level functionality. There are the algorithms which usually do very high level things, but internally they get their hands dirty with `std::vector`s sometimes forgetting error propagation or masking while dealing with partial bins each their own way. Having a good set of standard arithmetic operations between the mid-level data types (histograms) would help a lot here.

Reply:
Yes, basically you would want to run something like `std::transform` on an algorithm. I had once started implementing this, see https://github.com/mantidproject/mantid/pull/15465, but have not developed the approach any further since, so I would not recommend using it more at this point.

Reply:
Some users will consider this a feature instead of a bug. Sometimes we just need to say it's your foot.

Reply:
I agree with @quantumsteve, we cannot take away the ability of users to get access and manipulate the raw data - it limits their scope for playing with reduction/analysis.

Reply:
I think this is not meant to take away that functionality completely. For example, if we let user modify `Y` directly, they might change its size without also changing the length of `X` and `E` (this was possible before adding the `Histogram` related types which now prevent this, but think of similar cases).
Instead we should simply allow access to the histogram, but not the internals of histogram *on the workspace interface*. If users really want to modify `Y` directly they would need to do that via the interface of the histogram and then use for example `ws.setHistogram(mydata)`.